### ISSUE TITLE
deploy-minion.sh bug with 0 sized files

### ISSUE THREAD

**Issue by sajal** [turbobytes/pulse#24]
Minion was not starting up for an agent host. They inspected the SD card in PC and found minion binary, tar files and sig files to be of 0 bytes. My guess is something got corrupted, and our bash script does not deal with the files existing where they should be but are of wrong sizes. Perhaps curl should download files to temporary location and move them on success.

1. Figure out how it happened. Try to fix root cause
2. Try to reproduce. Make the bash script stronger by re-downloading if this happens. Perhaps one quick fix is to delete the `current` and `latest` files on issues.

This particular issue would have likely fixed itself when we deploy new version of minion since that triggers re-downloading.

**Comment by coolparadox** [comment]
A few tests in my computer shows that `curl -o` doesn't mess with output file in case of common network errors. This makes me agree with your suspicion of (filesystem?) corruption, which can be hard to reproduce.

Does `deploy-minion.sh` in the affected system make use of busybox shell or applet (eg. sha256sum)?

**Comment by coolparadox** [comment]
Branch `deploy_minion` contains an updated `deploy-minion.sh` that tries to safeguard against downloaded files with zero lenght, as well as validate and extract minion executable on every loop as another safeguard against file corruption.

PS: I have tested it in my computer only (amd64 gentoo linux os).

**Comment by sajal** [comment]
> Does `deploy-minion.sh` in the affected system make use of busybox shell or applet (eg. sha256sum)?

It mostly runs in bash. Particularly in this case.

So the key difference is you are doing a `rm -f current` on validation failure? 

Why do you do a `rm -f current` (line 166) each time after minion exits? By design when no valid certificate is present [minion will exit](https://github.com/turbobytes/pulse/blob/master/utils/minion.go#L174) and restart the loop. Removing the `current` file will cause the loop to download the files all over again. This will cause very high bandwidth usage.

**Comment by coolparadox** [comment]
Agreed; during coding I naively assumed that the sole reason minion exits is due to a crash. Fixing.

I asked about busybox because in my experience its applets (eg. sha256sum) eventually don't conform to their real counterparts on corner cases. `busybox sha256sum -c FILE`, in particular, exits with success code when `FILE` is empty.

As for detection of empty files during validation, the key difference is using `-s` flag instead of `-f` for detecting if files exist _and_ have sizes greater than zero.
